hsm-gateway-applications/go/hsm-sample.go

```diff
@@ -0,0 +1,200 @@
+
+	fmt.Printf("Submit Transaction: CreateAsset, creates new asset with ID, Color, Size, Owner and AppraisedValue arguments \n")
+
+	_, err := contract.SubmitTransaction("CreateAsset", assetId, "yellow", "5", "Tom", "1300")
+	if err != nil {
+		panic(fmt.Errorf("failed to submit transaction: %w", err))
+	}
+
+	fmt.Printf("*** Transaction committed successfully\n")
+
+	fmt.Printf("Evaluate Transaction: ReadAsset, function returns asset attributes\n")
+
+	evaluateResult, err := contract.EvaluateTransaction("ReadAsset", assetId)
+	if err != nil {
+		panic(fmt.Errorf("failed to evaluate transaction: %w", err))
+	}
+	result := formatJSON(evaluateResult)
+
+	fmt.Printf("*** Result:%s\n", result)
+}
+
+// newGrpcConnection creates a gRPC connection to the Gateway server.
+func newGrpcConnection() *grpc.ClientConn {
+	certificate, err := loadCertificate(tlsCertPath)
+	if err != nil {
+		panic(fmt.Errorf("failed to obtain commit status: %w", err))
+	}
+
+	certPool := x509.NewCertPool()
+	certPool.AddCert(certificate)
+	transportCredentials := credentials.NewClientTLSFromCert(certPool, "peer0.org1.example.com")
+
+	connection, err := grpc.Dial(peerEndpoint, grpc.WithTransportCredentials(transportCredentials))
+	if err != nil {
+		panic(fmt.Errorf("failed to evaluate transaction: %w", err))
+	}
+
+	return connection
+}
+
+// newIdentity creates a client identity for this Gateway connection using an X.509 certificate.
+func newIdentity(certificatePEM []byte) *identity.X509Identity {
+	cert, err := identity.CertificateFromPEM(certificatePEM)
+	if err != nil {
+		panic(err)
+	}
+	id, err := identity.NewX509Identity(mspID, cert)
+	if err != nil {
+		panic(err)
+	}
+
+	return id
+}
+
+// newHSMSign creates a function that generates a digital signature from a message digest using a private key.
+func newHSMSign(h *identity.HSMSignerFactory, certPEM []byte) (identity.Sign, identity.HSMSignClose) {
+	opt := identity.HSMSignerOptions{
+		Label:      "ForFabric",
+		Pin:        "98765432",
+		Identifier: string(certPEM),
+	}
+
+	sign, close, err := h.NewHSMSigner(opt)
+	if err != nil {
+		panic(err)
+	}
+
+	return sign, close
+}
+
+func loadCertificate(filename string) (*x509.Certificate, error) {
+	certificatePEM, err := ioutil.ReadFile(filename) //#nosec G304
+	if err != nil {
+		return nil, err
+	}
+
+	return identity.CertificateFromPEM(certificatePEM)
+}
+
+func getSKI(certPEM []byte) []byte {
+	block, _ := pem.Decode(certPEM)
+
+	x590cert, _ := x509.ParseCertificate(block.Bytes)
+	pk := x590cert.PublicKey
+
+	return skiForKey(pk.(*ecdsa.PublicKey))
+}
+
+func skiForKey(pk *ecdsa.PublicKey) []byte {
+	ski := sha256.Sum256(elliptic.Marshal(pk.Curve, pk.X, pk.Y))
+	return ski[:]
+}
+
+func findSoftHSMLibrary() string {
+
+	libraryLocations := []string{
+		"/usr/lib/softhsm/libsofthsm2.so",
+		"/usr/lib/x86_64-linux-gnu/softhsm/libsofthsm2.so",
+		"/usr/local/lib/softhsm/libsofthsm2.so",
+		"/usr/lib/libacsp-pkcs11.so",
+	}
+
```

Comment:
Needs to check PKCS11_LIB environment variable in case library is in a non-standard location:

```go
	pkcs11lib := os.Getenv("PKCS11_LIB")
	if pkcs11lib != "" {
		libraryLocations = append(libraryLocations, pkcs11lib)
	}
```